resource/aws_rds_cluster_parameter_group and resource/aws_db_parameter_group: Restore ability to change parameter values

camlow325 opened the pull request:
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Closes #11846
Closes #13551

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
* resource/aws_rds_cluster_parameter_group and resource/aws_db_parameter_group: Restore ability to change parameter values ([#11846](https://github.com/terraform-providers/terraform-provider-aws/issues/11846))
```

In #11540, support was added to the rds parameter group resources for resetting parameter values to AWS defaults when parameters are removed from config. A side-effect of these changes, however, was that parameters which remain in the config but whose values have been changed would also be reset to AWS defaults. This commit restores the ability for parameter values to be updated in the config while retaining the ability for parameters being removed from the config to be reset to AWS defaults.

Output from acceptance testing:

```
$  make testacc TEST=./aws TESTARGS='-run=TestAccAWSDBClusterParameterGroup_'
...
--- PASS: TestAccAWSDBClusterParameterGroup_disappears (18.01s)
--- PASS: TestAccAWSDBClusterParameterGroup_namePrefix (23.28s)
--- PASS: TestAccAWSDBClusterParameterGroup_only (23.40s)
--- PASS: TestAccAWSDBClusterParameterGroup_generatedName (23.42s)
--- PASS: TestAccAWSDBClusterParameterGroup_generatedName_Parameter (23.77s)
--- PASS: TestAccAWSDBClusterParameterGroup_namePrefix_Parameter (23.79s)
--- PASS: TestAccAWSDBClusterParameterGroup_withApplyMethod (23.89s)
--- PASS: TestAccAWSDBClusterParameterGroup_basic (86.64s)
```

```
$ make testacc TEST=./aws TESTARGS='-run=TestAccAWSDBParameterGroup_'
...
--- PASS: TestAccAWSDBParameterGroup_Disappears (18.18s)
--- PASS: TestAccAWSDBParameterGroup_generatedName (22.40s)
--- PASS: TestAccAWSDBParameterGroup_namePrefix (22.69s)
--- PASS: TestAccAWSDBParameterGroup_Only (23.40s)
--- PASS: TestAccAWSDBParameterGroup_MatchDefault (24.19s)
--- PASS: TestAccAWSDBParameterGroup_withApplyMethod (25.23s)
--- PASS: TestAccAWSDBParameterGroup_limit (45.08s)
--- PASS: TestAccAWSDBParameterGroup_basic (75.38s)
```

deiga commented:
@lachlancooper Could this PR get a review? This is a very annoying bug

lachlancooper commented:
> @lachlancooper Could this PR get a review? This is a very annoying bug

Agreed, the bug has a rather high potential for disastrous impact on production systems. This fix looks sensible to me, but I'm not a maintainer. The contributing docs say a maintainer will review it, but not how to request a review.

@bflad: would you mind taking a look at this (since you reviewed the original PR)?

mikkoc commented:
Any news on this? We're looking at adding parameter groups, but I'm very hesitant now given what I read.

Olliepop commented:
> Any news on this? We're looking at adding parameter groups, but I'm very hesitant now given what I read.

I recommend holding off on adding parameter groups until this is merged. Parameter values are reset to the default value on each apply, no matter what you specify. Until this PR is merged you must update the parameter value after applying a Terraform plan within the RDS console or with aws-cli.

The effect on production systems can be disastrous if you have a large number of parameter values which are not the RDS default.

dooferlad commented on aws/resource_aws_db_parameter_group.go:
```diff
@@ -290,8 +290,26 @@ func resourceAwsDbParameterGroupUpdate(d *schema.ResourceData, meta interface{})
 		}
 
+		// Build a map with names from the new parameter set
+		newNames := make(map[string]bool, ns.Len())
```
Shouldn't we move the removed / reset logic into the above modify loop so there is one place where a parameter change is managed? That way we can:

```
for , parameter := range parameters {
  if parameter in ns and not in os {
   add
  } else if parameter in ns and in os {
   modify
 } else if paremeter not in ns and in os {
  delete / reset to default
 }
```

The way we have the code now, and after this change, acts on modified values. The bug is that `parameters, err := expandParameters(ns.Difference(os).List())` generates the same set of values as `resetParameters, err := expandParameters(os.Difference(ns).List())`.

dooferlad replied:
Sorry, I was wrong, this is a set difference. Order matters - ns.Difference(os) is the values in ns that are not in os.

camlow325 replied:
Yeah, the parameters that have been removed shouldn't be caught by the `parameters, err := expandParameters(ns.Difference(os).List())` call which precedes the block which processes modified parameters since none of the removed parameters are in `ns`. As @lachlancooper explained really well in https://github.com/terraform-providers/terraform-provider-aws/issues/11846#issuecomment-586777896, the `resetParameters, err := expandParameters(os.Difference(ns).List())` call would catch any parameters whose name or other attributes (including value) appear in `os` but not in `ns`. This leads to parameters which have the same name in `os` and `ns` but different values getting unintentionally reset, which is what this PR intends to fix.

dooferlad commented on aws/resource_aws_db_parameter_group.go:
```diff
@@ -277,8 +277,26 @@ func resourceAwsDbParameterGroupUpdate(d *schema.ResourceData, meta interface{})
 			}
 		}
 
+		// Build a map with names from the new parameter set
```
I would be tempted to go with:

```go
		toRemove := map[string]*rds.Parameter{}

		for _, p := range expandParameters(os.List()) {
			toRemove[*p.ParameterName] = p
		}

		for _, p := range expandParameters(ns.List()) {
			delete(toRemove, *p.ParameterName)
		}

		// Reset parameters that have been removed
		var resetParameters []*rds.Parameter
		for _, v := range toRemove {
			resetParameters = append(resetParameters, v)
		}
```

It is a bit more obvious and lets `expandParameters` care about the format of the interface it receives.

camlow325 replied:
Thanks for this writeup, @dooferlad. This does seem like an improvement. If @anGie44 agrees, I'd be happy to swap in your implementation.

Ray-B commented:
Would be good to see this out as soon as is reasonable. This bug is a show-stopper for us, and I suspect many others who work in larger environments.

Without this fix, it effectively bricks our ability to make configuration changes to databases under RDS, particularily if you have limitations on your ability to make manual configuration changes to production environments / have complex deployment workflows, limitations on rollback, etc.

wgorski commented:
Any news on this? We are already using parameter groups on production environments and this bug makes it very difficult to apply any changes.

anGie44 commented on aws/resource_aws_rds_cluster_parameter_group_test.go:
```diff
@@ -176,6 +176,25 @@ func TestAccAWSDBClusterParameterGroup_basic(t *testing.T) {
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckAWSDBClusterParameterGroupExists(resourceName, &v),
+					testAccCheckAWSDBClusterParameterGroupAttributes(&v, parameterGroupName),
+					resource.TestCheckResourceAttr(
+						resourceName, "parameter.3127312348.name", "character_set_results"),
+					resource.TestCheckResourceAttr(
+						resourceName, "parameter.3127312348.value", "ascii"),
+					resource.TestCheckResourceAttr(
+						resourceName, "parameter.422535354.name", "character_set_server"),
+					resource.TestCheckResourceAttr(
+						resourceName, "parameter.422535354.value", "ascii"),
+					resource.TestCheckResourceAttr(
+						resourceName, "parameter.2478663599.name", "character_set_client"),
+					resource.TestCheckResourceAttr(
+						resourceName, "parameter.2478663599.value", "utf8"),
+				),
```
these Set indices unfortunately are no longer supported in acceptance testing since the recent upgrade to V2 of the terraform-plugin-sdk, but they can be easily updated to use the new convention:
```go
tfawsresource.TestCheckTypeSetElemNestedAttrs(resourceName, "parameter.*", map[string]string{
    "name": "character_set_results",
    "value": "ascii"", 
}),
```
where the attribute that is a Set is provided with an `.*` suffix like `parameter.*` and the map[string]string provided corresponds to the key/value pairs expected

camlow325 commented on aws/resource_aws_rds_cluster_parameter_group_test.go:
```diff
@@ -176,6 +176,25 @@ func TestAccAWSDBClusterParameterGroup_basic(t *testing.T) {
+					resource.TestCheckResourceAttr(
+						resourceName, "parameter.2478663599.name", "character_set_client"),
+					resource.TestCheckResourceAttr(
+						resourceName, "parameter.2478663599.value", "utf8"),
```
Thanks, I made those updates.

camlow325 commented on aws/resource_aws_db_parameter_group_test.go:
```diff
@@ -163,6 +163,25 @@ func TestAccAWSDBParameterGroup_basic(t *testing.T) {
+					resource.TestCheckResourceAttr(
+						resourceName, "parameter.2478663599.name", "character_set_client"),
+					resource.TestCheckResourceAttr(
+						resourceName, "parameter.2478663599.value", "utf8"),
```
Thanks, I made those updates.

camlow325 commented:
Thanks for all of the review comments, @anGie44.

> Overall, I would just recommend creating a new test with each of the new configs to test the new "update" behavior so we ensure we are not introducing a regression and we can attempt to keep the basic test as one that should only test for the minimum required fields, though understandably over time it has expanded quite a bit.

Sorry, just saw this comment after I reworked my previous test updates to use the `TestCheckTypeSetElemNestedAttrs` helper.`Update` had seemed like a part of `Basic` functionality for the resource but I agree that this test has a fair amount of other assertions that it's doing already. I can split this out into a new test.
 
> With regard to the code additions in both the db_parameter_group and rds_cluster_parameter_group resources, since the debug logs show that the ordering of the `Update` behavior transitions from `Modifying` then to `Reset` of the parameters, we could instead flip around that existing logic and that should allow a user to modify a param as intended without resetting soon after. This wouldn't require additional logic then what is already present in our codebase, just a shift in the ordering, but if you have any additional feedback or questions please let me know!

This would be less turmoil than what I had added for the PR. @lachlancooper mentioned in https://github.com/terraform-providers/terraform-provider-aws/issues/11846#issuecomment-586777896:

> Assuming we want to keep the functionality of resetting parameters that are being removed, one extremely simple fix would be to swap the order of the modify and reset operations so that modify always "wins". That's still not ideal as the operations would happen non-atomically, meaning every changing parameter would get the wrong value for a brief period.

I agree with this comment and this is the reason why I had gone with the approach of ensuring that parameters that have changed are only updated once during an apply. I worry that by just flipping the logic that the initial reset of a parameter to its default value rather than the intended new value could have negative repercussions - especially for "apply immediately" parameters or if a network issue were to happen between the "reset" and "modify" that could indefinitely leave one or more parameters unintentionally reset to their default values.

camlow325 commented:
> Overall, I would just recommend creating a new test with each of the new configs to test the new "update" behavior so we ensure we are not introducing a regression and we can attempt to keep the basic test as one that should only test for the minimum required fields, though understandably over time it has expanded quite a bit.

In 31e7b8f, I moved the new `TestStep` entries out into new test funcs - `TestAccAWSDBParameterGroup_updateParameters` `TestAccAWSDBClusterParameterGroup_updateParameters` - with separate configs for each.

anGie44 commented:
> by just flipping the logic that the initial reset of a parameter to its default value rather than the intended new value could have negative repercussions

I think this is a great point @camlow325 ! and definitely well illustrated in @lachlancooper's comment. I'd go ahead with @dooferlad's https://github.com/terraform-providers/terraform-provider-aws/pull/12112#discussion_r454467536 . I would just nitpick on the use of `*p.ParameterName` , to avoid any unwanted panics (not very likely from the API response, but we have had cases in the past where struct fields can be `nil`) we can use the AWS SDK Go method `aws.StringValue()` to safely get the pointer's value.

camlow325 commented on aws/resource_aws_rds_cluster_parameter_group.go:
```diff
@@ -221,16 +221,33 @@ func resourceAwsRDSClusterParameterGroupUpdate(d *schema.ResourceData, meta inte
+		for _, v := range toRemove {
+			resetParameters = append(resetParameters, v)
+		}
+		if len(resetParameters) > 0 {
```
The changes from here on down in this file just rename the `parameters` variable to `resetParameters` in order to avoid conflicting with the previously defined `parameters` variable and make this code look more like the code in the `resource_aws_db_parameter_group.go` file.

camlow325 commented:
> I'd go ahead with @dooferlad's #12112 (comment) . I would just nitpick on the use of *p.ParameterName , to avoid any unwanted panics (not very likely from the API response, but we have had cases in the past where struct fields can be nil) we can use the AWS SDK Go method aws.StringValue() to safely get the pointer's value.

I swapped in the new implementation in 1296450. For the `p.ParameterName` references, I just wrapped them with some `nil` checks to avoid adding an entry to or referencing an entry in the map for a `nil` value. That seemed a little better than just using a potential empty string key from `aws.StringValue` with the map. In any event, a `nil` value for `p.ParameterValue` would seem to be a sign of something bad/unintended.

anGie44 commented on aws/resource_aws_db_parameter_group.go:
```diff
@@ -277,8 +277,25 @@ func resourceAwsDbParameterGroupUpdate(d *schema.ResourceData, meta interface{})
 
+		toRemove := map[string]*rds.Parameter{}
+
+		for _, p := range expandParameters(os.List()) {
+			if p.ParameterName != nil {
+				toRemove[*p.ParameterName] = p
+			}
+		}
+
+		for _, p := range expandParameters(ns.List()) {
+			if p.ParameterName != nil {
+				delete(toRemove, *p.ParameterName)
+			}
+		}
```
nice 👍 I forgot since technically we're expanding user input here (in the rds cluster resource) and the "name"  field within the `parameter` block is `Required`, we shouldn't ever come across a `nil` (as well i see the `exandParameters` actually skips for cases "name" is an empty string) but i prefer erring on the side of caution when it comes to pointers.

jamesmharvey commented:
@anGie44 I know it's a long shot, but is there any chance of getting this released as a pointfix in a v2.70.0? There are a significant number of breaking changes in v3.0.0, and we're still working through them.
